opm/simulators/wells/WellStateFullyImplicitBlackoil.hpp

```diff
@@ -1007,29 +1007,67 @@ namespace Opm
         }
 
         template<class Comm>
-        void communicateGroupRates(const Comm& comm) {
-            // sum over all nodes
-            for (auto& x : injection_group_rein_rates) {
-                comm.sum(x.second.data(), x.second.size());
-            }
+        void communicateGroupRates(const Comm& comm)
+        {
+            // Doing this separately since it is a single double, not a vector.
             for (auto& x : injection_group_vrep_rates) {
                 x.second = comm.sum(x.second);
             }
-            for (auto& x : production_group_reduction_rates) {
-                comm.sum(x.second.data(), x.second.size());
-            }
-            for (auto& x : injection_group_reduction_rates) {
-                comm.sum(x.second.data(), x.second.size());
-            }
-            for (auto& x : injection_group_reservoir_rates) {
-                comm.sum(x.second.data(), x.second.size());
-            }
-            for (auto& x : production_group_rates) {
-                comm.sum(x.second.data(), x.second.size());
-            }
-            for (auto& x : well_rates) {
-                comm.sum(x.second.data(), x.second.size());
-            }
+
+            // Create a function that calls some function
+            // for all the individual data items to simplify
+            // the further code.
+            auto forAllGroupData = [&](auto& func) {
+                for (auto& x : injection_group_rein_rates) {
+                    func(x.second);
+                }
+                for (auto& x : production_group_reduction_rates) {
+                    func(x.second);
+                }
+                for (auto& x : injection_group_reduction_rates) {
+                    func(x.second);
+                }
+                for (auto& x : injection_group_reservoir_rates) {
+                    func(x.second);
+                }
+                for (auto& x : production_group_rates) {
+                    func(x.second);
+                }
+                for (auto& x : well_rates) {
+                    func(x.second);
+                }
+            };
+
+            // Compute the size of the data.
+            std::size_t sz = 0;
+            auto computeSize = [&sz](const auto& v) {
+                sz += v.size();
+            };
+            forAllGroupData(computeSize);
+
+            // Make a vector and collect all data into it.
+            std::vector<double> data(sz);
+            std::size_t pos = 0;
+            auto collect = [&data, &pos](const auto& v) {
+                for (const auto& x : v) {
+                    data[pos++] = x;
+                }
+            };
+            forAllGroupData(collect);
+            assert(pos == s);
```

Comment:
sz?

Reply:
Yes, I changed the name and did not notice until I built in debug.